File: python/src/wt/cleanup.py
from enum import Enum
from pathlib import Path
from typing import Dict
from wt.worktree_manager import WorktreeManager


class CleanupMode(Enum):
    AUTO = "auto"
    MANUAL = "manual"
    KEEP_ON_ERROR = "keep-on-error"


class CleanupOrchestrator:
    def __init__(self, worktree_manager: WorktreeManager):
        self.worktree_manager = worktree_manager

    def cleanup_worktree(
        self,
        task_id: str,
        worktree_dir: Path,
        mode: CleanupMode,
        task_failed: bool
    ):
        if mode == CleanupMode.MANUAL:
            return

        if mode == CleanupMode.KEEP_ON_ERROR and task_failed:
            return

        if self.worktree_manager.worktree_exists(task_id, worktree_dir):
            self.worktree_manager.remove_worktree(task_id, worktree_dir)

    def cleanup_all(
        self,
        worktree_dir: Path,
        mode: CleanupMode,
        task_statuses: Dict[str, bool]
    ):
        for task_id, failed in task_statuses.items():
            self.cleanup_worktree(task_id, worktree_dir, mode, failed)

File: src/cleanup.rs
use anyhow::Result;
use crate::worktree_manager::WorktreeManager;

#[derive(Debug, Clone, Copy)]
pub enum CleanupMode {
    Auto,
    Manual,
    KeepOnError,
}

impl CleanupMode {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "auto" => Some(CleanupMode::Auto),
            "manual" => Some(CleanupMode::Manual),
            "keep-on-error" => Some(CleanupMode::KeepOnError),
            _ => None,
        }
    }
}

pub struct CleanupOrchestrator<'a> {
    worktree_manager: &'a WorktreeManager,
}

impl<'a> CleanupOrchestrator<'a> {
    pub fn new(worktree_manager: &'a WorktreeManager) -> Self {
        Self { worktree_manager }
    }

    pub fn cleanup_worktree(
        &self,
        task_id: &str,
        mode: CleanupMode,
        task_failed: bool,
    ) -> Result<()> {
        match mode {
            CleanupMode::Manual => Ok(()),
            CleanupMode::KeepOnError if task_failed => Ok(()),
            _ => {
                if self.worktree_manager.get_worktree_info(task_id)?.is_some() {
                    self.worktree_manager.remove_worktree(task_id)?;
                }
                Ok(())
            }
        }
    }
}
